### DISCUSSION TITLE
ERROR on proc 0: Substitution for illegal variable NSTEPS (../input.cpp:651)

### DISCUSSION THREAD

**Discussion by SHANG-Ming-Hui** [deepmodeling/deepmd-kit#3021] [Q&A]
Dear all,
Can you kindly give me some guidence for addressing the error encountered during the **model_devi** step ? 
The error message is ** ERROR on proc 0: Substitution for illegal variable NSTEPS (../input.cpp:651)
Last command: run             ${NSTEPS} upto **.
The number of processes in a task is the number of times this error appeares.

My environment:  **deepmd-kit-2.2.4-gpu** version coupled with **LAMMPS-stable_2Aug2023**
                            **DP-GEN-0.10.4** and **dpdata-0.2.4** were both installed.

Model can be trained using the "dpgen run" flow. While the "model_devi" could not be performed due to the error raised as bellow:
_**ERROR on proc 0: Substitution for illegal variable NSTEPS (../input.cpp:651)
Last command: run             ${NSTEPS} upto**_

The corresponding log file was attached.
The parameter.json for **model_devi**  step is set as:
*******************************************************
    "model_devi_dt": 0.001,
    "model_devi_skip": 0,
    "model_devi_f_trust_lo": 0.05,
    "model_devi_f_trust_hi": 0.15,
    "model_devi_e_trust_lo": 10000000000.0,
    "model_devi_e_trust_hi": 10000000000.0,
    "model_devi_clean_traj": true,
    "model_devi_jobs": [
      {"sys_idx": [0], "temps": [500, 380, 300], "press": [1.0], "trj_freq": 10, "nsteps": 2000, "ensemble": "nvt", "_idx": "00"}
********************************************************

The machine.json for **model_devi** step is set as:
********************************************************
"machine": {
                "context_type": "local",
                "batch_type": "slurm",
                "local_root": "./",
                "remote_root": "xxxxxxxxxxxxxxxxxxxxxx(path)/gputrain-2/tmp/md",
                "_comment": "that's all"
            },
            "resources": {
                "number_node": 1,
                "cpu_per_node": 48,
                "gpu_per_node": 4,
                "queue_name": "gpu",
                "strategy":{"if_cuda_multi_devices":true},
                "custom_flags": [
                    "#SBATCH -J LMP",
                    "#SBATCH -N 1",
                    "#SBATCH --ntasks-per-node=4",
                    "#SBATCH --ntasks-per-socket=2",
                    "#SBATCH --cpus-per-task=10",
                    "#SBATCH -t 24:00:00",
                    "#SBATCH --mem=32G"
                ],
                "source_list": [
                    "/public/home/smh/software/lmp.env"
                ],
                "module_list": [],
                "_time_limit": "",
                "group_size": 5
            },
            "command": "mpirun -np 4 ~/software/lammps-stable_2Aug2023/bin/lmp_mpi -sf gpu -pk gpu 1 tpa 8"
********************************************************

Best wishes
12/1, 2023
[model_devi.log](https://github.com/deepmodeling/deepmd-kit/files/13526093/model_devi.log)

**Comment by njzjz** [comment]
The error is caused by the LAMMPS script and should be not related to DeePMD-kit.

**Reply by SHANG-Ming-Hui** [reply]
Dear Jinzhe,
thanks so much for your kind reply.
Can you please kindly give me some guide to address the problem ?
The lammps has been recomipled for several times, however, the error remains.
Using the same param.json, **dpgen run** can work well and complete the flow using the deepmd-kit-cpu version with a lower training efficiency.

**Reply by njzjz** [reply]
Perhaps `-sf gpu -pk gpu 1 tpa 8` should be removed. But the error message given by LAMMPS is a misunderstanding

**Reply by SHANG-Ming-Hui** [reply]
Hi Jinzhe, thanks so much for help.
Acctually, i have tried to run lmp_mpi simply as **"mpirun -np 4 lmp_mpi -i input.lammps -v restart 0"** on the shell terminal. However, the error changed like below:
*********************************************************************************
2023-12-04 16:09:14.297177: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1532] Created device /job:localhost/replica:0/task:0/device:GPU:3 with 36318 MB memory:  -> device: 3, name: NVIDIA A100-PCIE-40GB, pci bus id: 0000:a4:00.0, compute capability: 8.0
2023-12-04 16:09:14.297348: I tensorflow/core/common_runtime/process_util.cc:146] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.
2023-12-04 16:09:14.326110: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:354] MLIR V1 optimization pass is not enabled
2023-12-04 16:09:14.329183: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:354] MLIR V1 optimization pass is not enabled
2023-12-04 16:09:14.329433: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:354] MLIR V1 optimization pass is not enabled
2023-12-04 16:09:14.332480: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:354] MLIR V1 optimization pass is not enabled
INVALID_ARGUMENT: Tensor spin_attr/ntypes_spin:0, specified in either feed_devices or fetch_devices was not found in the Graph
INVALID_ARGUMENT: Tensor spin_attr/ntypes_spin:0, specified in either feed_devices or fetch_devices was not found in the Graph
INVALID_ARGUMENT: Tensor spin_attr/ntypes_spin:0, specified in either feed_devices or fetch_devices was not found in the Graph
INVALID_ARGUMENT: Tensor spin_attr/ntypes_spin:0, specified in either feed_devices or fetch_devices was not found in the Graph
lmp_mpi: symbol lookup error: lmp_mpi: undefined symbol: DP_DeepPotIsAParamNAll
lmp_mpi: symbol lookup error: lmp_mpi: undefined symbol: DP_DeepPotIsAParamNAll
lmp_mpi: symbol lookup error: lmp_mpi: undefined symbol: DP_DeepPotIsAParamNAll
lmp_mpi: symbol lookup error: lmp_mpi: undefined symbol: DP_DeepPotIsAParamNAll
*********************************************************************************
I'm now carefully checking the compiled lammps by recomputing the general examples from LAMMPS package.
The research has currently stalled due to this difficulty. 
More guidance and suggestions are welcome, and sincerely thank you all in advance.

**Reply by njzjz** [reply]
This is a version mismatching. `DP_DeepPotIsAParamNAll` is a new symbol, and an old DeePMD-kit library might be loaded.

`export LD_DEBUG=libs` will help to find which library is actually loaded instead of the correct one.

**Reply by SHANG-Ming-Hui** [reply]
Hi, Jinzhe, thanks so much.
Currently, i found the model_deviation can be performed by manually changing the "**run ${NSTEPS} upto**" as "**run 2000 upto**" in the last line of **input.lammps** generated within the "dpgen run" work flow. Thus the lmp_mpi seems to be compiled successfuly. 
Does it imply that the interface between DeepMD-kit and lammps was incorrectly build ?
Thanks again.

**Reply by njzjz** [reply]
I don't understand why `${NSTEPS}` didn't work.

It's also unclear why it worked with the CPU version.

**Reply by njzjz** [reply]
You may post the entire LAMMPS input file.

**Reply by SHANG-Ming-Hui** [reply]
**The inputs for LAMMPS running with error:**
**********************************************************************************************************************************
`**plugin load libdeepmd_lmp.sovariable        NSTEPS          equal 1000**  #####   _Normally, a line break is required. However, there is no line break in the current file, which leads to an error in variable setting._
variable        THERMO_FREQ     equal 10
variable        DUMP_FREQ       equal 10
variable        TEMP            equal 500.000000
variable        PRES            equal -1.000000
variable        TAU_T           equal 0.100000
variable        TAU_P           equal 0.500000

units           metal
boundary        p p p
atom_style      atomic

neighbor        1.0 bin

box          tilt large
if "${restart} > 0" then "read_restart dpgen.restart.*" else "read_data conf.lmp"
change_box   all triclinic
mass            1 133.000000
mass            2 119.000000
mass            3 127.000000
pair_style      deepmd ../graph.000.pb ../graph.001.pb ../graph.002.pb ../graph.003.pb  out_freq ${THERMO_FREQ} out_file model_devi.out
pair_coeff      * *

thermo_style    custom step temp pe ke etotal press vol lx ly lz xy xz yz
thermo          ${THERMO_FREQ}
dump            1 all custom ${DUMP_FREQ} traj/*.lammpstrj id type x y z fx fy fz
restart         10000 dpgen.restart

if "${restart} == 0" then "velocity        all create ${TEMP} 786184"
fix             1 all nvt temp ${TEMP} ${TEMP} ${TAU_T}

timestep        0.001000
**run             ${NSTEPS} upto**`
**********************************************************************************************************************************

**The inputs for LAMMPS works well:**
`**plugin load libdeepmd_lmp.sovariable        NSTEPS          equal 1000**  #####   _Normally, a line break is required. However, there is no line break in the current file, which leads to an error in variable setting._
variable        THERMO_FREQ     equal 10
variable        DUMP_FREQ       equal 10
variable        TEMP            equal 500.000000
variable        PRES            equal -1.000000
variable        TAU_T           equal 0.100000
variable        TAU_P           equal 0.500000

units           metal
boundary        p p p
atom_style      atomic

neighbor        1.0 bin

box          tilt large
if "${restart} > 0" then "read_restart dpgen.restart.*" else "read_data conf.lmp"
change_box   all triclinic
mass            1 133.000000
mass            2 119.000000
mass            3 127.000000
pair_style      deepmd ../graph.000.pb ../graph.001.pb ../graph.002.pb ../graph.003.pb  out_freq ${THERMO_FREQ} out_file model_devi.out
pair_coeff      * *

thermo_style    custom step temp pe ke etotal press vol lx ly lz xy xz yz
thermo          ${THERMO_FREQ}
dump            1 all custom ${DUMP_FREQ} traj/*.lammpstrj id type x y z fx fy fz
restart         10000 dpgen.restart

if "${restart} == 0" then "velocity        all create ${TEMP} 786184"
fix             1 all nvt temp ${TEMP} ${TEMP} ${TAU_T}

timestep        0.001000
**run             2000 upto**`
**********************************************************************************************************************************

**For the CPU-version, LAMMPS is not loaded as the GPU-version. So the LAMMPS task can be perfomed normally by DeepMD-kit-CPU, while failed in the GPU version.
Perhaps,  the problem can be resolved by modifying the code or something else. I'm now trying to do this.**

**Comment by SHANG-Ming-Hui** [comment]


**Reply by SHANG-Ming-Hui** [reply]
Presently, the problem can been solved by adding a line break at the end of the statement "**ret = "plugin load libdeepmd_lmp.so"**" within the dpgen/generator/lib/lammps.py-plugin. 
Which should be changed as:
********************************************************************************************************************************************
ret = "plugin load libdeepmd_lmp.so **\n**"   **#### adding a line break.**
********************************************************************************************************************************************
